aten/src/ATen/BatchedTensorImpl.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <bitset>
+
+#include <ATen/ArrayRef.h>
+#include <ATen/SmallVector.h>
+#include <ATen/Tensor.h>
+
+namespace at {
+
+// We assume this in a few other places in the codebase,
+// but there isn't a centralized definition.
+constexpr int64_t kVmapMaxTensorDims = 64;
+
+struct BatchDim {
+  BatchDim(int64_t dim, int64_t level) : dim_(dim), level_(level) {}
+  int64_t dim() const {
+    return dim_;
+  }
+  int64_t level() const {
+    return level_;
+  }
+ private:
+  int64_t dim_;
+  int64_t level_;
+};
+
+using BatchDims = SmallVector<BatchDim, 5>;
+using BatchDimsRef = ArrayRef<BatchDim>;
+
+struct TORCH_API BatchedTensorImpl : public c10::TensorImpl {
+  explicit BatchedTensorImpl(Tensor value, BatchDims bdims);
+
+  BatchDimsRef bdims() const { return bdims_; }
+  const Tensor& value() const { return value_; };
+
+  int64_t actualDim(int64_t dim, bool wrap_dim = true) const;
+
+  // Override a bunch of methods inherited from TensorImpl.
+  bool is_contiguous(at::MemoryFormat memory_format=at::MemoryFormat::Contiguous) const override;
+  IntArrayRef strides() const override;
+  int64_t stride(int64_t d) const override;
+  void set_size(int64_t dim, int64_t new_size) override;
+  void set_stride(int64_t dim, int64_t new_stride) override;
+  void set_storage_offset(int64_t storage_offset) override;
+  bool has_storage() const override;
+  const Storage& storage() const override;
+  int64_t storage_offset() const override;
+
+ private:
+  Tensor value_;
+  BatchDims bdims_;
+};
+
+inline bool isBatched(const Tensor& tensor) {
+  return tensor.unsafeGetTensorImpl()->key_set().has(DispatchKey::Vmap);
+}
+
+inline BatchedTensorImpl* getBatched(Tensor tensor) {
+  return static_cast<BatchedTensorImpl*>(tensor.unsafeGetTensorImpl());
```

Comment:
`unsafeGetTensorImpl`, why is unsafe? sounds... unsafe

Reply:
Unsafe because it returns the underlying TensorImpl. Users of our C++ API for instance should not be using it and the word "unsafe" deters them.

Reply:
Also, in my experience, it even gives you the impl if the tensor is constant. So you could modify a constant Tensor.

Reply:
It might be more in line with the prevailing naming if this one is also called getBatched. In fact, this is indeed unsafe, because it is not valid to call this on a non-batched tensor.

Reply:
In this PR, you use getBatched in conjunction with isBatched. So an even safer form would be maybeGetBatched() that returns a nullptr if the result is not a batched tensor impl.

Reply:
I renamed `getBatched` to `unsafeGetBatched` and I created a `maybeGetBatched` variant (thanks for the suggestions!).

I have some downstream code that relies on `unsafeGetBatched` so I'd like to keep it around but we should be able to delete it in the future